### PR TITLE
TST: expand tests for ExtensionArray setitem with nullable arrays

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Follow-up on https://github.com/pandas-dev/pandas/pull/31484 to add some more test cases (eg also test the validation done in the `check_array_indexer` (wrong length, missing values), as well for nullable integer arrays), similarly to the tests I added for `__getitem__`.

cc @charlesdong1991

### CODE REVIEW

File: pandas/tests/extension/base/setitem.py [hunk 1]
```diff
@@ -93,6 +92,91 @@ def test_setitem_iloc_scalar_multiple_homogoneous(self, data):
...
+    def test_setitem_mask_boolean_array(self, data, box_in_series):
+        # GH 31446 - nullable boolean mask
+        arr = data[:5].copy()
+        expected = arr.take([0, 0, 0, 3, 4])
+        if box_in_series:
+            arr = pd.Series(arr)
+            expected = pd.Series(expected)
+        mask = pd.array([True, True, True, False, False], dtype="boolean")
+        arr[mask] = data[0]
+        self.assert_equal(expected, arr)
```
Comment (jbrockmendel) [inline pandas/tests/extension/base/setitem.py]: can this test and the previous one be combined+parametrized?

Comment (jorisvandenbossche) [reply pandas/tests/extension/base/setitem.py]: yep, done

File: pandas/tests/extension/base/setitem.py [hunk 2]
```diff
@@ -93,6 +92,86 @@ def test_setitem_iloc_scalar_multiple_homogoneous(self, data):
...
+
+    def test_setitem_mask_raises(self, data, box_in_series):
+        # wrong length
+        mask = np.array([True, False])
+
+        if box_in_series:
+            data = pd.Series(data)
+
+        with pytest.raises(IndexError):
```
Comment (charlesdong1991) [inline pandas/tests/extension/base/setitem.py]: not sure if an error message should be matched here and below?

**Review by charlesdong1991 — COMMENTED** [review]
awesome!! Thanks for adding tests for it! @jorisvandenbossche

File: pandas/tests/extension/base/setitem.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/setitem.py]: Added part of the message

File: pandas/tests/extension/base/setitem.py [hunk 3]
```diff
@@ -93,6 +92,86 @@ def test_setitem_iloc_scalar_multiple_homogoneous(self, data):
...
+        ids=["list", "integer-array"],
+    )
+    def test_setitem_integer_with_missing_raises(self, data, idx):
+        arr = data.copy()
+
+        msg = "Cannot index with an integer indexer containing NA values"
+        with pytest.raises(ValueError, match=msg):
+            arr[idx] = arr[0]
+
+        # TODO this raises KeyError about labels not found (it tries label-based)
```
Comment (jreback) [inline pandas/tests/extension/base/setitem.py]: xfail instead?

**Review by jreback — CHANGES_REQUESTED** [review]
small comment, pls rebase  as I think we merged something that might impact this (or maybe its open still, the string setitem change).

File: pandas/tests/extension/base/setitem.py (hunk 3 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/setitem.py]: Added xfail, it made it a bit more complicated, but also discovered that it actually worked for series with array indexer! (only the list is not working)

File: pandas/tests/extension/base/setitem.py [hunk 4]
```diff
@@ -93,6 +93,90 @@ def test_setitem_iloc_scalar_multiple_homogoneous(self, data):
...
+            expected = pd.Series(expected)
+
+        arr[idx] = arr[0]
+        self.assert_equal(arr, expected)
+
+    @pytest.mark.parametrize(
+        "idx, box_in_series",
+        [
+            ([0, 1, 2, pd.NA], False),
+            pytest.param([0, 1, 2, pd.NA], True, marks=pytest.mark.xfail),
```
Comment (jreback) [inline pandas/tests/extension/base/setitem.py]: you can add a message for the xfail (do we have an issue for this?)

File: pandas/tests/extension/base/setitem.py [hunk 5]
```diff
@@ -93,6 +93,90 @@ def test_setitem_iloc_scalar_multiple_homogoneous(self, data):
...
+            "Cannot mask with a boolean indexer containing NA values|"
+            "cannot mask with array containing NA / NaN values"
+        )
+        with pytest.raises(ValueError, match=msg):
+            data[mask] = data[0]
+
+    @pytest.mark.parametrize(
+        "idx",
+        [[0, 1, 2], pd.array([0, 1, 2], dtype="Int64"), np.array([0, 1, 2])],
+        ids=["list", "integer-array", "numpy-array"],
```
Comment (jreback) [inline pandas/tests/extension/base/setitem.py]: labels should be: extension-integer-array, numpy-integer-array to be more informative

File: pandas/tests/extension/base/setitem.py [hunk 6]
```diff
@@ -93,6 +93,90 @@ def test_setitem_iloc_scalar_multiple_homogoneous(self, data):
         df.iloc[10, 1] = data[1]
         assert df.loc[10, "B"] == data[1]
 
+    @pytest.mark.parametrize(
+        "mask",
+        [
+            np.array([True, True, True, False, False]),
+            pd.array([True, True, True, False, False], dtype="boolean"),
+        ],
+        ids=["numpy-array", "boolean-array"],
```
Comment (jreback) [inline pandas/tests/extension/base/setitem.py]: numpy-boolean-array, extension-boolean-array

File: pandas/tests/extension/base/setitem.py [hunk 7]
```diff
@@ -93,6 +93,90 @@ def test_setitem_iloc_scalar_multiple_homogoneous(self, data):
         df.iloc[10, 1] = data[1]
         assert df.loc[10, "B"] == data[1]
 
+    @pytest.mark.parametrize(
```
Comment (jreback) [inline pandas/tests/extension/base/setitem.py]: think about adding a new class here; these are specifically for setitem_mask (could be a followon)

File: pandas/tests/extension/base/setitem.py [hunk 8]
```diff
@@ -93,6 +93,90 @@ def test_setitem_iloc_scalar_multiple_homogoneous(self, data):
...
+        "mask",
+        [
+            np.array([True, True, True, False, False]),
+            pd.array([True, True, True, False, False], dtype="boolean"),
```
Comment (jreback) [inline pandas/tests/extension/base/setitem.py]: add a list, maybe even add Index-object dtype of bools

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
needs rebase

File: pandas/tests/extension/base/setitem.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/setitem.py]: > you can add a message for the xfail (do we have an issue for this?)

Opened an issue here: https://github.com/pandas-dev/pandas/issues/31948

File: pandas/tests/extension/base/setitem.py (hunk 7 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/setitem.py]: > think about adding a new class here; these are specifically for setitem_mask (could be a followon)

In general, changing names is annoying for downstream consumers of those tests (this are the base extension tests), so not worth here I think

File: pandas/tests/extension/base/setitem.py (hunk 8 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/setitem.py]: >  add Index-object dtype of bools

That is explicitly not supported in general (and that the validation raises an error for that is already tested elsewhere)

File: pandas/tests/extension/base/setitem.py (hunk 6 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/setitem.py]: The test name indicates it is about masking, so I think "numpy-array" should be sufficient

File: pandas/tests/extension/base/setitem.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/tests/extension/base/setitem.py]: And added that issue to the xfail message

File: pandas/tests/extension/base/setitem.py (hunk 7 above)
Comment (jreback) [reply pandas/tests/extension/base/setitem.py]: ok, I don't buy your argument, these are experimental features where we actually to expand test coverage. its going to happen (but ok for now)

**Review by jreback — APPROVED** [review]
lgtm. @TomAugspurger

**Review by TomAugspurger — APPROVED** [review]

**Pull request merged**